### DISCUSSION TITLE
Use of `contextlib.contextmanager` with `Depends`

### DISCUSSION THREAD

**Discussion by eseglem** [fastapi/fastapi#8955] [Questions]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Description

Based on [A dependency with yield and try](https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-with-yield/#a-dependency-with-yield-and-try) I wanted to try and reuse the same code to manage sessions in other parts of my app (tests, management scripts, setup, etc.). So based on my reading of the [Tip](https://github.com/tiangolo/fastapi/blame/master/docs/en/docs/tutorial/dependencies/dependencies-with-yield.md#L214-L226) at the bottom, I tried adding `@contextlib.contextmanager` to the function. Which prevents `Depends(get_db)` from working correctly. 

The Tip says you don't *have* to use the decorators, and that you *shouldn't*. To me, this implies it is possible to use the decorator, but discouraged. Unfortunately, it appears it is not possible. I would guess this functionality is probably working as intended, and the docs can be updated to be more clear. Perhaps just remove reference to `@contextlib.contextmanager`? 

I did dig into the code to see if there was something obvious that I could do to make this work, but I was not able to come up with anything yet. 

### Example

Below is a standalone snippet to show the current functionality. It remains true for async generators as well. 

```Python
from contextlib import contextmanager
from typing import Generator
from fastapi import FastAPI, Depends

class DBSession:
    def query(self, text: str) -> str:
        return text

    def close(self) -> None:
        pass

def get_session_without() -> Generator[DBSession, None, None]:
    session = DBSession()
    try:
        yield session
    finally:
        session.close()

@contextmanager
def get_session_with() -> Generator[DBSession, None, None]:
    session = DBSession()
    try:
        yield session
    finally:
        session.close()

app = FastAPI()

@app.get("/without")
async def route_without_session(
    session: DBSession = Depends(get_session_without),
) -> str:
    return session.query("blah")

@app.get("/with")
async def route_with_session(
    session: DBSession = Depends(get_session_with),
) -> str:
    return session.query("blah")

if __name__ == "__main__":
    from fastapi.testclient import TestClient

    client = TestClient(app)
    print(client.get("/without"))
    print(client.get("/with"))
```

This results in the following:
```
<Response [200]>
Traceback (most recent call last):
  File "mvp.py", line 54, in <module>
    print(client.get("/with"))
...
  File "mvp.py", line 46, in route_with_session
    return session.query("blah")
AttributeError: '_GeneratorContextManager' object has no attribute 'query'
```

### Environment

* OS: Ubuntu-20.04
* FastAPI Version: 0.61.1
* Python version: 3.8.5

### Additional context

The workaround is something like this: 

```python
@contextmanager
def get_session_with() -> Generator[DBSession, None, None]:
    session = DBSession()
    try:
        yield session
    finally:
        session.close()

def get_session_without() -> Generator[DBSession, None, None]:
    with get_session_with() as session:
        yield session
```

Where you `Depends(get_session_without)` on your routes, and `with get_session_with() as session:` anywhere else you want to use it. Which is totally OK. 

### Unrelated to the issue

Happy to dig through issues, and see how I can contribute, but I am curious is it also helpful to hang out on gitter and try to assist people on there? Is that perhaps something to add to the "After submitting this, I commit to one of:" section at the top? I have been doing it anyways, and will keep doing so when I have time either way. Just wondering.

**Comment by jan-golda** [comment]
I would like to dump this one as I think this would be a great improvement to FastAPI.

Let's say I have a simplified service like this:
```python
class FooService:
    def __init__(self):
        self._conn = None

    def __enter__(self):
        self._conn = BarConnection()  # init connection
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        self._conn.close()  # close connection

    def do_something():
        self._conn.do_something()  # use connection
```

If support for context managers was added I could simply use it this way:
```python
@app.get("/endpoint")
def endpoint(foo_service: FooService = Depends()):
    foo_service.do_something()
```

But unfortunately, now I have to use a [workaround from documentation](https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-with-yield/#using-context-managers-in-dependencies-with-yield):
```python
def get_foo_service():
    with FooService() as service:
        yield service

@app.get("/endpoint")
def endpoint(foo_service: FooService = Depends(get_foo_service)):
    foo_service.do_something()
```

I know it's essentially an aesthetic change, but for some people stuff like this matters.

I would be happy to contribute with the needed changes if only I could get some pointers from the maintainers on how they would see this implemented.

**Comment by a-recknagel** [comment]
I don't know why noone has said it yet, but it's not possible<sup>*</sup> to know if a function was decorated with `contextlib.contextmanager` without actually calling it. On inspection, it looks just like the undecorated function, plus `__wrapped__`, which is not conclusive. For that reason, the [initial feature request](https://github.com/tiangolo/fastapi/issues/49#issuecomment-466726099) included a proposal where `Depends` would receive a new keyword argument `ctx_mgr` that could be set to `True` in order to tell fastapi that calling the passed function would produce a context manager.  

But that option did not make it to the final implementation. Pure speculation on my part, but increasing the footprint of a class as central as `Depends` was probably not deemed worth the payoff. Any new kwarg needs to be learned by users, and since `asynccontextmanager`s are just as elusive, you'd need `async_ctx_mgr` as well.

Adding support for a class-based context manager as @jan-golda would be easy, but was probably not included in order to be consistent.

---
<sup>*</sup> It is possible by inspection of either the function's code object, or the associated source code. Either of these approaches are hard to maintain though, and don't seem worth the effort when there is a straight-forward workaround.

**Reply by notpushkin** [reply]
Should be possible for async functions, which [are already called when resolving dependencies][1]:

```py
@asynccontextmanager
async def get_session():
    yield 42

get_session.__aenter__  # => AttributeError: 'function' object has no attribute '__aenter__'
# but...
called = get_session()  # => <contextlib._AsyncGeneratorContextManager at ...>

if hasattr(called, "__aenter__"):  # => True
    solved = await stack.enter_async_context(called)
else:
    solved = await called
```

[1]: https://github.com/tiangolo/fastapi/blob/1bf5e7a10e33d6f65e0f701ec271785fbc9b463c/fastapi/dependencies/utils.py#L583-L586


I haven't investigated how hard would it be for regular context managers, but in principle it might work too.

Update: I've drafted a PR implementing this: https://github.com/tiangolo/fastapi/pull/10353
